libs/patchwheel/patch.go

```diff
@@ -0,0 +1,280 @@
+	scanner := bufio.NewScanner(r)
+	var newLines []string
+	for scanner.Scan() {
+		line := scanner.Text()
+		if strings.TrimSpace(line) == "" {
+			continue
+		}
+		parts := strings.Split(line, ",")
+		if len(parts) < 3 {
+			// If the line doesn't have enough parts, preserve it as-is
+			newLines = append(newLines, line)
+			continue
+		}
+		origPath := parts[0]
+		if strings.HasPrefix(origPath, oldDistInfoPrefix) {
+			parts[0] = newDistInfoPrefix + origPath[len(oldDistInfoPrefix):]
+		}
+		// For the METADATA file entry, update hash and size.
+		if strings.HasSuffix(parts[0], "METADATA") {
+			parts[1] = "sha256=" + metadataHash
+			parts[2] = strconv.Itoa(metadataSize)
+		}
+		newLines = append(newLines, strings.Join(parts, ","))
+	}
+	if err := scanner.Err(); err != nil {
+		return nil, err
+	}
+	return []byte(strings.Join(newLines, "\n") + "\n"), nil
+}
+
+// PatchWheel patches a Python wheel file by updating its version in METADATA and RECORD.
+// It returns the path to the new wheel.
+// The version is updated according to the following rules:
+//   - if there is an existing part after + it is dropped
+//   - append +<mtime of the original wheel> to version
+//
+// The function is idempotent: repeated calls with the same input will produce the same output.
+// If the target wheel already exists, it returns the path to the existing wheel without redoing the patching.
+func PatchWheel(ctx context.Context, path, outputDir string) (string, error) {
+	fileInfo, err := os.Stat(path)
+	if err != nil {
+		return "", err
+	}
+	wheelMtime := fileInfo.ModTime().UTC()
+
+	filename := filepath.Base(path)
+	wheelInfo, err := ParseWheelFilename(filename)
+	if err != nil {
+		return "", err
+	}
+
+	newVersion, newFilename := CalculateNewVersion(wheelInfo, wheelMtime)
+	outpath := filepath.Join(outputDir, newFilename)
+
+	if _, err := os.Stat(outpath); err == nil {
+		// Target wheel already exists, return its path
+		return outpath, nil
+	}
+
+	tmpFilename := outpath + fmt.Sprintf(".tmp%d", os.Getpid())
+
+	needRemoval := true
+
+	defer func() {
+		if needRemoval {
+			_ = os.Remove(tmpFilename)
+		}
+	}()
+
+	r, err := zip.OpenReader(path)
+	if err != nil {
+		return "", err
+	}
+	defer r.Close()
+
+	metadataFile, recordFile, oldDistInfoPrefix := readMetadataAndRecord(r)
+	if metadataFile == nil {
+		return "", fmt.Errorf("wheel %s missing METADATA file", path)
+	}
+
+	if recordFile == nil {
+		return "", fmt.Errorf("wheel %s missing RECORD file", path)
+	}
+
+	metadataReader, err := metadataFile.Open()
+	if err != nil {
+		return "", err
+	}
+	defer metadataReader.Close()
+
+	newMetadata, err := patchMetadata(metadataReader, wheelInfo.Version, newVersion)
+	if err != nil {
+		return "", err
+	}
+
+	h := sha256.New()
+	h.Write(newMetadata)
+	metadataHash := base64.URLEncoding.WithPadding(base64.NoPadding).EncodeToString(h.Sum(nil))
+	metadataSize := len(newMetadata)
```

Comment:
I believe this should be file size of METADATA, correct? If so, does the file size differ when written to disk on Windows because of `\r\n` new line characters?

Reply:
> I believe this should be file size of METADATA, correct?
Correct.

> If so, does the file size differ when written to disk on Windows because of \r\n new line characters?
It is \n on all platforms when inside zip. I don't think it's modified upon unpacking, but even if it is, this is up to pip/uv.